### PR TITLE
Adds a timeout to all external commands. Closes #1821

### LINKED ISSUE #1821: Forked tool processes should timeout
**Describe the bug**

All forked processes should implement a tool timeout so that the repo server does not hang.

### PR DESCRIPTION
**Pull request by alexec**

### CODE REVIEW

File: cmd/argocd/commands/project_role.go [hunk 1]
```diff
@@ -7,7 +7,7 @@ import (
 	"strconv"
 	"text/tabwriter"
 
-	timeutil "github.com/argoproj/pkg/time"
```
Comment (alexec) [inline cmd/argocd/commands/project_role.go]: needs to be reverted once change merged

File: util/kube/ctl.go [hunk 2]
```diff
@@ -247,33 +246,12 @@ func runKubectl(kubeconfigPath string, namespace string, args []string, manifest
 		cmdArgs = append(cmdArgs, "--dry-run")
 	}
 	cmd := exec.Command("kubectl", cmdArgs...)
-	log.Info(cmd.Args)
-	if log.IsLevelEnabled(log.DebugLevel) {
```
Comment (alexec) [inline util/kube/ctl.go]: no need for this

File: reposerver/repository/repository.go [hunk 3]
```diff
@@ -533,7 +533,7 @@ func runCommand(command v1alpha1.Command, path string, env []string) (string, er
 	cmd := exec.Command(command.Command[0], append(command.Command[1:], command.Args...)...)
 	cmd.Env = env
 	cmd.Dir = path
-	return argoexec.RunCommandExt(cmd)
+	return argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline reposerver/repository/repository.go]: this is used by config management plugins

File: util/git/client.go [hunk 4]
```diff
@@ -294,18 +295,5 @@ func (m *nativeGitClient) runCmdOutput(cmd *exec.Cmd) (string, error) {
...
-		exErr, ok := err.(*exec.ExitError)
-		if ok {
-			errOutput := strings.Split(string(exErr.Stderr), "\n")[0]
-			log.Debug(errOutput)
-			return string(out), fmt.Errorf("'%s' failed: %v", strings.Join(cmd.Args, " "), errOutput)
-		}
-		return string(out), fmt.Errorf("'%s' failed: %v", strings.Join(cmd.Args, " "), err)
-	}
-	return string(out), nil
+	return argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/git/client.go]: this is used for git

File: util/helm/helm.go [hunk 5]
```diff
@@ -224,21 +225,8 @@ func (h *helm) helmCmdExt(args []string, logFormat func(string) string) (string,
...
-		}
-		errOutput := string(exErr.Stderr)
-		log.Errorf("`%s` failed: %s", cmdStr, errOutput)
-		return "", fmt.Errorf(strings.TrimSpace(errOutput))
-	}
-	out := string(outBytes)
-	log.Debug(out)
-	return out, nil
+
+	return argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/helm/helm.go]: helm, obvs

File: util/ksonnet/ksonnet.go [hunk 6]
```diff
@@ -98,21 +98,7 @@ func (k *ksonnetApp) ksCmd(args ...string) (string, error) {
...
-			return "", err
-		}
-		errOutput := string(exErr.Stderr)
-		log.Errorf("`%s` failed: %s", cmdStr, errOutput)
-		return "", fmt.Errorf(strings.TrimSpace(errOutput))
-	}
-	out := string(outBytes)
-	log.Debug(out)
-	return out, nil
+	return argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/ksonnet/ksonnet.go]: ksonnet

File: util/kube/ctl.go [hunk 7]
```diff
@@ -247,33 +246,12 @@ func runKubectl(kubeconfigPath string, namespace string, args []string, manifest
...
-		}
-		redactedBytes, err := json.Marshal(redacted)
-		if err != nil {
-			return "", err
-		}
-		log.Debug(string(redactedBytes))
-	}
 	cmd.Stdin = bytes.NewReader(manifestBytes)
-	out, err := cmd.Output()
+	out, err := argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/kube/ctl.go]: kubectl

File: util/kustomize/kustomize.go [hunk 8]
```diff
@@ -120,7 +120,7 @@ func (k *kustomize) Build(opts *v1alpha1.ApplicationSourceKustomize) ([]*unstruc
 			args = append(args, arg)
 			cmd := exec.Command(commandName, args...)
 			cmd.Dir = k.path
-			_, err := argoexec.RunCommandExt(cmd)
+			_, err := argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/kustomize/kustomize.go]: kustomize

File: util/kustomize/kustomize.go [hunk 9]
```diff
@@ -82,7 +82,7 @@ func (k *kustomize) Build(opts *v1alpha1.ApplicationSourceKustomize) ([]*unstruc
 				for _, override := range opts.ImageTags {
 					cmd := exec.Command(commandName, "edit", "set", "imagetag", fmt.Sprintf("%s:%s", override.Name, override.Value))
 					cmd.Dir = k.path
-					_, err := argoexec.RunCommandExt(cmd)
+					_, err := argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/kustomize/kustomize.go]: kustomize

File: util/kustomize/kustomize.go [hunk 10]
```diff
@@ -69,7 +69,7 @@ func (k *kustomize) Build(opts *v1alpha1.ApplicationSourceKustomize) ([]*unstruc
 		if opts.NamePrefix != "" {
 			cmd := exec.Command(commandName, "edit", "set", "nameprefix", opts.NamePrefix)
 			cmd.Dir = k.path
-			_, err := argoexec.RunCommandExt(cmd)
+			_, err := argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/kustomize/kustomize.go]: kustomize

File: util/kustomize/kustomize.go [hunk 11]
```diff
@@ -135,7 +135,7 @@ func (k *kustomize) Build(opts *v1alpha1.ApplicationSourceKustomize) ([]*unstruc
 	}
 	defer func() { _ = closer.Close() }()
 	cmd.Env = append(cmd.Env, environ...)
-	out, err := argoexec.RunCommandExt(cmd)
+	out, err := argoexec.RunCommandExt(cmd, argoexec.CmdOpts{})
```
Comment (alexec) [inline util/kustomize/kustomize.go]: kustomize

**Comment by alexec** [comment]
We need to decide on default timeouts for 

* Config Management Plugins
* Jsonnet
* Helm 
* Git
* Kustomize

Currently, this is never. 

We could have settings for this

**Comment by jessesuen** [comment]
Let's have a globally defined argoproj/pkg/exec.CmdOpts used throughout reposerver for all config tools and git commands. We can configure via an environment variable just so that we have an escape hatch for users who cannot live with our hard-coded default. There's no need to make an Argo CD setting for it.

My initial reaction is that this should be 90 seconds. Jsonnet/Ksonnet is notorious for taking a long time to compute.

**Comment by alexec** [comment]
We've all decided that an envvar would be best at the same time.

File: util/config/exec.go [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"os"
+	"time"
+
+	"github.com/argoproj/pkg/exec"
+)
+
+func timeout() time.Duration {
+	duration, err := time.ParseDuration(os.Getenv("ARGOCD_EXEC_TIMEOUT"))
```
Comment (jessesuen) [inline util/config/exec.go]: Environment variables never really change once the program starts, I think we can calculate a private `time.Duration` value as part of an init() function to optimize all of the unnecessary calls to `time.ParseDuration`, `os.Getenv` `timeout()`, which will only ever get set once.

File: util/kube/ctl.go (hunk 2 above)
Comment (jessesuen) [reply util/kube/ctl.go]: We still need this. This is printing the object that we are about to apply. That's not part of CLI command or outputs.

**Review by jessesuen — CHANGES_REQUESTED** [review]

File: util/config/exec.go (hunk 12 above)
Comment (alexec) [reply util/config/exec.go]: kk

Comment (alexec) [reply util/config/exec.go]: um, how do I test it?

**Review by alexmt — APPROVED** [review]
LGTM

**Review by jessesuen — APPROVED** [review]

**Pull request merged**